controller_adapters/database.go

```diff
@@ -82,3 +84,102 @@ func (d *Database) GetProfileResolvers() ndb_api.ProfileResolvers {
 	return profileResolvers
 
 }
+
+// Returns basic details about the Time Machine if provided in the
+// underlying database, else returns defaults like:
+// TM Name: db_instance_name_TM
+// TM Description: Time Machine for db_instance_name
+// SLA Name: NONE
+func (d *Database) GetTMDetails() (tmName, tmDescription, slaName string) {
+	tmInfo := d.Spec.Instance.TMInfo
+
+	tmName = tmInfo.Name
+	tmDescription = tmInfo.Description
+	slaName = tmInfo.SLAName
+
+	if tmName == "" {
+		tmName = d.GetDBInstanceName() + "_TM"
+	}
+	if tmDescription == "" {
+		tmDescription = "Time Machine for " + d.GetDBInstanceName()
+	}
+	if slaName == "" {
+		slaName = common.SLA_NAME_NONE
+	}
+
+	return
+}
+
+// Returns a schedule struct for the time machine.
+func (d *Database) GetTMSchedule() (schedule ndb_api.Schedule, err error) {
+	tmInfo := d.Spec.Instance.TMInfo
+	hhmmssDaily := strings.Split(tmInfo.DailySnapshotTime, ":")
+	if len(hhmmssDaily) != 3 {
+		err = errors.New("invalid DailySnapshotTime, use the (24-hour) hh:mm:ss format")
+		return
```

Comment:
if err != nil, we should return an invalid `schedule` (nil value instead of a zero-initialized object). 
if one return value is non-nil (in this example `err`), the other should/cannot also be non-nil (`schedule` in this example). I am not sure of the implications of this while writing UTs because while mocking the response we are expecting a zero-initialized object at the moment. Not sure if we can mock/expect `nil`

Reply:
With the current implementation, it is assumed that the client/caller always checks for the err before making use of the returned `schedule` object. But yes, I see your point. Though I cannot return a nil schedule/struct in golang, I can change the return type to be a pointer to the struct and then return a nil pointer. Sounds good?

Reply:
Okay, so I changed the function (and the interface definition) to return a pointer to the schedule struct. Now our test will need to check for the returned pointer value - which will not be known to us, a reference to the schedule is created within the execution scope of the function and returned and we cannot guess/provide it to the test as an expected value. 

The only check we can perform is to check if the pointer is nil or not, this purpose is being served by the nil err check exactly, making our tests adding no value. Hence I feel returning empty/zero initialized objects with a non-nil err might be the correct way. We can discuss more, let me know how you feel about this. :)

Reply:
talked offline - returning a nil pointer to the struct will cause the operator to panic in case someone tries to reference it. we will leave the behavior as-is for now.